examples/apps/src/ble_nus_peripheral.rs

```diff
@@ -0,0 +1,129 @@
+
+// GATT Server
+#[gatt_server]
+struct Server {
+    nrf_uart: NrfUartService,
+}
+
+// NRF UART Service
+#[gatt_service(uuid = "6E400001-B5A3-F393-E0A9-E50E24DCCA9E")]
+struct NrfUartService {
+    #[characteristic(uuid = "6E400002-B5A3-F393-E0A9-E50E24DCCA9E", write)]
+    rx: Vec<u8, ATT_MTU>,
+
+    #[characteristic(uuid = "6E400003-B5A3-F393-E0A9-E50E24DCCA9E", notify)]
+    tx: Vec<u8, ATT_MTU>,
+}
+
+pub async fn run<C>(controller: C)
+where
+    C: Controller,
+{
+    let address = Address::random([0x41, 0x5A, 0xE3, 0x1E, 0x83, 0xE7]);
+    info!("Our address = {:?}", address);
+
+    let mut resources = Resources::new(PacketQos::None);
+    let (stack, peripheral, _, runner) = trouble_host::new(controller, &mut resources)
+        .set_random_address(address)
+        .build();
+
+    let server = Server::new_with_config(
+        stack,
+        GapConfig::Peripheral(PeripheralConfig {
+            name: "TrouBLE",
+            appearance: &appearance::GENERIC_UNKNOWN,
+        }),
+    );
+
+    info!("Starting advertising and GATT service");
+    let _ = join3(
+        ble_task(runner),
+        gatt_task(&server),
+        advertise_task(peripheral, &server),
+    )
+    .await;
+}
+
+async fn ble_task<C: Controller>(mut runner: Runner<'_, C>) -> Result<(), BleHostError<C::Error>> {
+    runner.run().await
+}
+
+async fn gatt_task<C: Controller>(server: &Server<'_, C>) {
+    loop {
+        match server.next().await {
+            Ok(GattEvent::Write { handle, connection: _ }) => {
+                let _ = server.get(handle, |value| {
+                    info!("[gatt] Write event on {:?}. Value written: {:?}", handle, value);
+                });
+            }
+            Ok(GattEvent::Read { handle, connection: _ }) => {
+                info!("[gatt] Read event on {:?}", handle);
+            }
+            Err(e) => {
+                error!("[gatt] Error processing GATT events: {:?}", e);
+            }
+        }
+    }
+}
+
+async fn advertise_task<C: Controller>(
+    mut peripheral: Peripheral<'_, C>,
+    server: &Server<'_, C>,
+) -> Result<(), BleHostError<C::Error>> {
+    let mut adv_data = [0; 31];
+    AdStructure::encode_slice(
+        &[
+            AdStructure::Flags(LE_GENERAL_DISCOVERABLE | BR_EDR_NOT_SUPPORTED),
+            AdStructure::CompleteLocalName(b"Trouble NRF UART (NUS) Service"),
+        ],
+        &mut adv_data[..],
+    )?;
+    loop {
+        info!("[adv] advertising");
+        let mut advertiser = peripheral
+            .advertise(
+                &Default::default(),
+                Advertisement::ConnectableScannableUndirected {
+                    adv_data: &adv_data[..],
+                    scan_data: &[],
+                },
+            )
+            .await?;
+        let conn = advertiser.accept().await?;
+
+        /* TODO: Implement "echo" and push rx bytes back to tx? */
+        // XXX: I have to manually "fix" the size from 123 -> 128
```

Comment:
@lulf Any clues what I'm doing wrong here? :)

Reply:
@plaes Can you explain a little more about the issue? Does something break if you make the `tx` array 123 bytes?

Reply:
If you have following on line 118, it asserts that 123 != 128 when sending the notification after someone has connected:
```rust
let mut tx = [0; ATT_MTU];
```

Reply:
Interesting...I'm guessing this is the asset in question: https://github.com/embassy-rs/trouble/blob/51d2b95a29b258b7b48f877cdf7af8a0ba54cd75/host/src/attribute.rs#L381

The issue you're facing is because your `tx` characteristic is of type `heapless::Vec<u8, 123>`, while the local variable `tx` is of type `[u8; 123]`. `heapless::Vec` has 4 bytes of metadata associated with it, giving 127 bytes which is probably being padded to 128 for alignment :)

Reply:
Following up on this, I'd definitely recommend using the same type throughout. With your code as it is currently written, line 124 is actually modifying the first byte in the `heapless::Vec`'s length field (which is of type `usize`). This will cause undefined behaviour if you were to try to use it as a `Vec` elsewhere.

On the whole I would be reluctant to use non-primitive types for characteristics as it assumes that the other end of the comms channel has an understanding of that type. Furthermore, the heapless implementation of Vec could change its memory structure in a future version, which would change the representation of the data in memory and cause further confusion.

TL;DR: in this instance I'd strongly recommend making your `tx` characteristic a byte array.

Reply:
Firstly, nice catch regarding the Vec vs byte array mismatch...

Though, using `heapless::Vec` hasn't been issue so far with nrf-softdevice library - https://github.com/lulf/watchful/blob/main/firmware/app/src/ble.rs

Reply:
I'm not particularly familiar with nrf-softdevice, but I believe that nrf-softdevice has logic to deal with characteristics defined as `Vec` type as variable-length characteristics. So on transmission it will only send `vector.len()` bytes, and on reception it will update the vector's length to the length of the received data.
At present (correct me if I'm wrong @lulf) TrouBLE doesn't have any of this logic and treats all characteristics as a simple block of data. So your `tx` characteristic will be sent as 128 bytes regardless of the current length of the vector.

I think support for variable-length characteristics is certainly something we want to add to TrouBLE though.

Reply:
This is where the assertion happens:
https://github.com/embassy-rs/trouble/blob/51d2b95a29b258b7b48f877cdf7af8a0ba54cd75/host/src/attribute.rs#L376-L383

Regarding nrf-softdevice, all the value conversions seem to go through `to_gatt()` and `from_gatt()` methods, which is provided by two traits: [`FixedGattValue`](https://github.com/embassy-rs/nrf-softdevice/blob/991d28e3c61687908acc53e748191eb59d00e27a/nrf-softdevice/src/ble/gatt_traits.rs#L12-L22) and [`GattValue`](https://github.com/embassy-rs/nrf-softdevice/blob/991d28e3c61687908acc53e748191eb59d00e27a/nrf-softdevice/src/ble/gatt_traits.rs#L24-L35).

GattValue can handle data with length smaller than attribute size and also implements support for Vec and String types via Heapless.

Reply:
Nice, I suspect we'll want to just implement the same, or possibly do so as part of the bt-hci crate.

Reply:
Yes, I'm in the process of bringing the FixedGattValue and GattValue traits into TrouBLE as part of my work on #136. The purpose of that work isn't to implement support for variable length characteristics, but it will provide the framework to be able to implement this in future.

Reply:
I've raised #149 to track the variable length characteristic implementation

Reply:
Agreed bringing over FixedGattValue and GattValue traits to trouble makes sense. There is also #133 which might also allow doing varlength without having to store it.